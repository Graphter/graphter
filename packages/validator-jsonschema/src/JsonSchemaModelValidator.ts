import JsonSchemaValidator from "./JsonSchemaValidator";
import { JsonSchemaValidatorOptions } from "./JsonSchemaValidatorOptions";
import {
  ModelConfig,
  ModelValidator,
  PropertyValidator,
  ValidationExecutionStage,
  ValidationResult
} from "microo-core";
import generateErrorMessage from "./generateErrorMessage";

export default class JsonSchemaModelValidator extends JsonSchemaValidator implements ModelValidator{
  constructor(options: JsonSchemaValidatorOptions) {
    super(options);
  }
  execute(stage: ValidationExecutionStage, config: ModelConfig, data: any): Promise<ValidationResult> {
    const valid = this._validateFn(data);
    if(valid){
      return Promise.resolve({
        valid: true
      });
    } else {
      const result: ValidationResult = {
        valid: false,
        errorMessage: generateErrorMessage(data, this._options, this._validateFn)
      }

      if(this._validateFn.errors){
        result.dataPaths = this._validateFn.errors.map(err => err.dataPath);
      }

      if(Array.isArray(this._options.errorDisplayMode)){
        result.errorDisplayMode = this._options.errorDisplayMode;
      } else if(this._options.errorDisplayMode !== undefined){
        result.errorDisplayMode = [ this._options.errorDisplayMode ];
      }

      return Promise.resolve(result);
    }
  }
}