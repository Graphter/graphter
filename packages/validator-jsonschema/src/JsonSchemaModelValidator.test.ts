import { ModelConfig, ValidationErrorDisplayMode, ValidationExecutionStage } from "microo-core";
import JsonSchemaModelValidator from "./JsonSchemaModelValidator";

describe(`JsonSchemaModelValidator`, () => {
  let modelConfig: ModelConfig;
  beforeEach(() => {
    modelConfig = {
      id: 'some-model',
      name: 'Some Model',
      properties: []
    }
  })
  it(`should return a valid response when executed with correct data`, async () => {
    const validator = new JsonSchemaModelValidator({
      schema: {
        minimum: 10,
        maximum: 12
      }
    });
    const result = await validator.execute(ValidationExecutionStage.CHANGE, modelConfig, 11);
    expect(result).not.toBeNull();
    expect(result.valid).toBe(true);
  });
  it(`should return an invalid response when executed with incorrect data`, async () => {
    const validator = new JsonSchemaModelValidator({
      schema: {
        minimum: 10,
        maximum: 12
      },
      error: `Must be at least 10 but no more than 12`
    });
    const result = await validator.execute(ValidationExecutionStage.CHANGE, modelConfig, 13);
    expect(result).not.toBeNull();
    expect(result.valid).toBe(false);
    expect(result.errorMessage).toBe(`Must be at least 10 but no more than 12`);
  });
  it(`should error when no schema is provided`, () => {
    expect(() => {
      // @ts-ignore
      new JsonSchemaModelValidator();
    }).toThrowErrorMatchingSnapshot();
  });
  it(`should set executeOn values correctly`, () => {
    const validator = new JsonSchemaModelValidator({
      schema: {
        maxLength: 5
      },
      executeOn: [
        ValidationExecutionStage.CHANGE
      ]
    });
    expect(validator.executeOn).not.toBeNull();
    expect(validator.executeOn).toHaveLength(1);
    expect(validator.executeOn[0]).toBe(ValidationExecutionStage.CHANGE);
  });
  it(`should return correct error display mode`, async () => {
    const validator = new JsonSchemaModelValidator({
      schema: {
        maxLength: 5
      },
      errorDisplayMode: ValidationErrorDisplayMode.MODAL
    });
    const result = await validator.execute(ValidationExecutionStage.CLIENT_UPDATE, modelConfig, 'Hello World!');
    expect(result.errorDisplayMode).not.toBeNull();
    expect(result.errorDisplayMode).toEqual([ ValidationErrorDisplayMode.MODAL ]);
  })
  it(`should default options correctly`, async () => {
    const validator = new JsonSchemaModelValidator({
      schema: {
        maxLength: 5
      }
    });
    expect(validator.executeOn).toMatchSnapshot();
    const result = await validator.execute(ValidationExecutionStage.CLIENT_UPDATE, modelConfig, 'Hello World!');
    expect(result.errorDisplayMode).toMatchSnapshot()
  });
});

export {};