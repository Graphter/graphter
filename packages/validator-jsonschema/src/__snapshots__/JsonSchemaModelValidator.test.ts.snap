// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonSchemaModelValidator should default options correctly 1`] = `
Array [
  "CLIENT_CREATE",
  "CLIENT_UPDATE",
  "SERVER_UPDATE",
  "SERVER_UPDATE",
]
`;

exports[`JsonSchemaModelValidator should default options correctly 2`] = `
Array [
  "INLINE",
  "SUMMARY",
]
`;

exports[`JsonSchemaModelValidator should error when no schema is provided 1`] = `"Schema option is required by JsonSchemaValidator"`;
